### PR TITLE
Documentation on graph visualization

### PR DESCRIPTION
**Pull request by iosonofabio**
Hi guys,

As promised I started drafting some docs on graph visualization via `igraph` (well, `python-igraph` I s'pose).

For now it describes a bit the graph layouts, and I am starting to tackle the plotting itself. So about that, it would be great to make a rough plan of action (beyond docs).

First of all:

https://stackoverflow.com/questions/3285193/how-to-change-backends-in-matplotlib-python

Second, I use ipython with matplotlib and `Qt5Agg` backend (it opens up interactive windows, quite convenient). There are a bunch of other interactive backends. How do we use igraph with those? The current docstrings only describe `cairo` or other static backends,but that seems limiting at best given we have the layout already. I usually just use igraph to get a layout and then do the plotting in a little function, e.g.:

```python
import igraph as ig
import matplotlib.pyplot as plt

g = ig.Graph.Full(3)
layout = g.layout('auto')

def plot_graph(g, layout, ax=None):
    x, y = list(zip(*layout.coords))
    if ax is None:
        fig, ax = plt.subplots()
    else:
       fig = None
    ax.scatter(x, y, s=30, color='black', zorder=10)
    for v1, v2 in g.get_edgelist():
        x1, y1, x2, y2 = x[v1], y[v1], x[v2], y[v2]
       ax.plot([x1, x2], [y1, y2], lw=2, color='black', zorder=9)
    if fig is not None:
        fig.tight_layout()
    return {'fig': fig, 'ax': ax}

plot_graph(g, layout)
```

It seems easy enough to do, is it already part of igraph or, if not, should I add it?

Any suggestions on how to write this second part of the doc page? Or should I first implement this kind of plotting function and then finish this doc PR?

Thanks!
Fabio

### CODE REVIEW

**Comment by ntamas** [comment]
Well, the plotting landscape has changed quite a lot since the time we've added plotting functionality to igraph -- right now there are so many options and backends to choose from that I don't really know what the best solution would be.

Back in the old days, Jupyter was in its infancy and so was Matplotlib, so I sort-of assumed that the plotting functionality in igraph has to be standalone because it's unclear who the "winner" will be in the long run (or whether there would be a winner at all). Cairo seemed to be an obvious choice, and the original idea was to support stuff like `plot(g)` in the Python prompt -- this would plot the graph to a temporary file, and open the built-in preferred image viewer of the platform with that temporary file. It sort-of-kind-of worked, apart from a few weird race conditions in certain platforms where the temporary file was removed _before_ the image viewer got a chance to open it. Right now it's still the officially supported way of plotting graphs if you don't have Jupyter.

In Jupyter, the case is a bit different because the `Graph` object has a `_repr_png_()` method that produces a PNG representation of the graph, which can then be shown by Jupyter. (We also have `_repr_svg_()`, but then there's the problem that you have to put each SVG in an `iframe`, otherwise the glyph IDs produced by Cairo would conflict if you have more than one plot in the same Jupyter workbook).

As for your Matplotlib solution, yes, it works as a quick-and-dirty solution, but it does not handle most of the visual attributes that the standard `plot()` method in igraph does. (I'm thinking about stuff like vertex and edge colors, edge weights, labels and so on). So, a long while ago I came up with a [hack that allows igraph graphs in a Matplotlib plot](https://stackoverflow.com/a/36154077/156771). I have no idea whether it still works now or it is outdated. The downside of this solution is that it needs the Cairo backend for Matplotlib, not the Agg backend.

All in all, I think that it would be great to have a solution in `python-igraph` that works nicely with Matplotlib (irrespectively of the chosen plotting backend), but it should also support the visual properties that the standard `plot()` function supports.

**Comment by ntamas** [comment]
As for the second graph: I think the two methods that should definitely be mentioned there are:

1. Simply calling `plot(graph)`, which plots the graph to a temporary file and shows it in the default image viewer of the platform.

2. Calling `plot(graph, "filename.png")` or `plot(graph, "filename.svg")`, which plots the graph to a persistent file, but does not open a window automatically.

**Comment by iosonofabio** [comment]
Thank you. Yes my code was just illustrating the concept. If we want to support matplotlib as a whole, then I'd be happy to make a PR for that that colors things properly.

Personally I think that'd be useful, even though some details (e.g. label positioning in a crowded graph) will be out of scope.

**Comment by szhorvat** [comment]
Just a tip: instead of simply listing layout names, mention in 1-2 words what they are, or what they are good for.

E.g., Sugiyama is for directed acyclic graphs, and for trees

Kamada-Kawai is sometimes described as a "srping layout" since it's based on a physical model of spring

Furchterman-Reingold is a "spring-electric layout", as it's based on a physical model of springs and electrostatic repulsion

graphopt and DrL are also force-directed

Davidson-Harel is based on simulated annealing optimization of a complex function that includes things such as edge crossings

**Comment by iosonofabio** [comment]
Awesome thank you. What is Sugiyama's advantage over other DAG/tree layouts?

**Comment by szhorvat** [comment]
The usual tree layouts work only on trees. Sugiyama works on any DAG. To draw DAGs nicely, one needs to use curved edges. I think this is what it does.

**Comment by szhorvat** [comment]
According to this page, Sugiyama was the first to develop this type of drawing https://en.wikipedia.org/wiki/Layered_graph_drawing

**Comment by iosonofabio** [comment]
Great, I'll add it, thank you!
Fabio

**Comment by szhorvat** [comment]
Some more tips for demonstrating the methods (maybe on a different, more details tutorial page):

For tree layouts, we can use `igraph_tree_game` to generate a random tree. I contributed this function some time ago. I think it's not yet exposed in Python, but it should be.

For DAG layouts (Sugiyama), first we generate an undirected Erdős–Rényi graph, then we convert it to directed with the new `ACYCLIC` method. This is already exposed in Python. This effectively generates a random DAG.

Random trees tend to look pretty good with most of the layouts implemented in igraph: http://szhorvat.net/mathematica/IGDocumentation/#gallery :-)

**Comment by iosonofabio** [comment]
Great ideas! I think that's better in the tutorial, like you said. I'll add a link to the tutorial in here and then we can edit that page separately.

**Comment by iosonofabio** [comment]
@ntamas in a Jupyter notebook, how does one specify to use the PNG representation to avoid the bug? I just opened a notebook and typed:

image

and I got an SVG (I think).

**Comment by ntamas** [comment]
Unfortunately I have no idea; the easiest is to remove the `_repr_svg_` method :) [This page on SO](https://stackoverflow.com/questions/45560556/what-is-the-precedence-order-of-jupyter-rich-representations) claims that all representations are generated by Jupyter (which is a bit unfortunate as we do lots of extra legwork for no reason then), and the most appropriate representation is chosen by the frontend, so in the end you have to look in the frontend configuration to force it to prefer PNG over SVG.

File: doc/source/visualisation.rst [hunk 1]
```diff
@@ -1,4 +1,145 @@
...
+
+and methods:
+
+- :meth:`Layout.boundaries` the rectangle with the extreme coordinates of the layout
+- :meth:`Layout.bounding_box` the boundary, but as an `igraph.drawing.utils.BoundingBox` (see below)
+- :meth:`Layout.centroid` the coordinates of the centroid of the graph layout
+
+Indexing and slicing can be performed and returns the coordinates of the requested vertices:
+
+>>> coords_subgraph = layout[:2]  # Coordinate of the first vertex
```
Comment (ntamas) [inline doc/source/visualisation.rst:40]: `layout[:2]` should return the coordinates of the first two vertices; `layout[0]` is the one that returns the coordinates of the first vertex.

File: doc/source/visualisation.rst [hunk 2]
```diff
@@ -1,4 +1,145 @@
...
+
+>>> ig.plot(g, 'myfile.pdf')
+
+.. note:: PNG is a raster image format while PDF, SVG, and Postscript are vector image formats. Choose one of the last three
+   formats if you are planning on refining the image with a vector image editor such as Inkscape or Illustrator.
+
+In both cases, you can add an argument `layout` to the `plot` function to specify a precomputed layout, e.g.:
+
+>>> layout = g.layout("kamada_kawai")
+>>> ig.plot(g, layout=layout)
```
Comment (ntamas) [inline doc/source/visualisation.rst:112]: It is also possible to just use the name of the layout algorithm in `ig.plot()`, e.g.: `ig.plot(g, layout="kamada_kawai")`. if you don't add a layout, igraph uses the dedicated `layout_auto()` function, which chooses between one of several possible layouts based on the number of vertices and edges.

File: doc/source/visualisation.rst (hunk 1 above)
Comment (iosonofabio) [reply doc/source/visualisation.rst:40]: fixed

File: doc/source/visualisation.rst (hunk 2 above)
Comment (iosonofabio) [reply doc/source/visualisation.rst:112]: Yep, I added that info now

**Comment by iosonofabio** [comment]
I added instructions for matplotlib plotting, that makes this PR conditional on #341

**Comment by ntamas** [comment]
Looks good to me! I'll merge this when we are ready with #341

**Pull request merged**